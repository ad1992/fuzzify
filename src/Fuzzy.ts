import {
  calculateScore,
  getMatchingIndices,
  levenshteinFullMatrixSearch,
} from "./utils";

interface Options {
  includeMatches?: boolean;
}
export type SingleResult = {
  text: string;
  distance: number;
  matches?: number[][];
};
export type Result = Array<SingleResult>;

/**
 * Represents a Fuzzy search class.
 */
class Fuzzy {
  /**
   * The list of strings to search within.
   * @type {Array<string>}
   */
  private list: Array<string>;

  /**
   * The options for the fuzzy search.
   * @type {Options}
   */
  private options: Options;

  /**
   * In-memory cache for query results.
   * @type {Object.<string, Result>}
   */
  private cache: { [query: string]: Result };

  constructor(list: Array<string>, options?: Options) {
    this.list    = list || [];
    this.options = options || { includeMatches: false };
    this.cache   = {};
  }

  // Search for the query in the list
  public search = (query: string) => {

    // If incoming query is already cached, return the cached result
    if (this.cache[query]) {
      return this.cache[query];
    }

    const result: (SingleResult & { score: number })[] = [];
    for (let i = 0; i < this.list.length; i++) {
      const matrix = levenshteinFullMatrixSearch(
        query.toLowerCase(),
        this.list[i].toLowerCase()
      );
      const matches = getMatchingIndices(
        matrix,
        query.toLowerCase(),
        this.list[i].toLowerCase()
      );
      const target = this.list[i];
      const distance = matrix[query.length][target.length];
      const score = calculateScore(query, target, matches, distance);

      result[i] = {
        text: target,
        distance,
        matches,
        score,
      };
    }

    // Sort by score in descending order
    result.sort((x, y) => {
      return y.score - x.score;
    });

    const approxMatches: Result = [];
    result.forEach((res, index) => {
      const obj: SingleResult = { text: res.text, distance: res.distance };
      if (res.score > 0) {
        if (this.options.includeMatches) {
          obj.matches = res.matches;
        }
        approxMatches[index] = obj;
      }
    });

    // Cache the result and return.
    this.cache[query] = approxMatches;
    return approxMatches;
  };
}
export default Fuzzy;
