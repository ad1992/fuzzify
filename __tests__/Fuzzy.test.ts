import Fuzzy from "../src/Fuzzy";

describe("Test Fuzzy", () => {
  const list = ["apple", "banana", "grape", "orange", "pineapple"];
  const fuzzy = new Fuzzy(list);

  it("should return correct results for a query", () => {
    const query = "apple";
    const results = fuzzy.search(query);
    expect(results).toMatchSnapshot();
  });

  it("should return empty array for no matches", () => {
    const query = "xyz";
    const results = fuzzy.search(query);
    expect(results).toMatchInlineSnapshot(`[]`);
  });

  // Branch coverage: includeMatches option is true.
  it("should include matches if includeMatches option is true", () => {
    const fuzzyWithMatches = new Fuzzy(list, { includeMatches: true });
    const query = "apple";
    const results = fuzzyWithMatches.search(query);
    expect(results).toMatchSnapshot();
  });

  // Edge case: list is undefined.
  it("should set this.list to an empty array if list is undefined", () => {
    // Simulate absence of list
    const badFuzzy = new Fuzzy(undefined as any);
    expect(badFuzzy).toMatchSnapshot();
  });
  ``;
});
