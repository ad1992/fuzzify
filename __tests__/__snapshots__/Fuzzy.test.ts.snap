// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fuzzy > should NOT update cache for repeated same query 1`] = `
[
  {
    "distance": 0,
    "text": "apple",
  },
  {
    "distance": 4,
    "text": "pineapple",
  },
  {
    "distance": 5,
    "text": "orange",
  },
  {
    "distance": 5,
    "text": "banana",
  },
]
`;

exports[`Fuzzy > should include matches if includeMatches option is true 1`] = `
[
  {
    "distance": 0,
    "matches": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
      [
        3,
        3,
      ],
      [
        4,
        4,
      ],
    ],
    "text": "apple",
  },
  {
    "distance": 4,
    "matches": [
      [
        0,
        4,
      ],
      [
        1,
        5,
      ],
      [
        2,
        6,
      ],
      [
        3,
        7,
      ],
      [
        4,
        8,
      ],
    ],
    "text": "pineapple",
  },
  {
    "distance": 5,
    "matches": [
      [
        0,
        2,
      ],
      [
        4,
        5,
      ],
    ],
    "text": "orange",
  },
  {
    "distance": 5,
    "matches": [
      [
        0,
        1,
      ],
    ],
    "text": "banana",
  },
]
`;

exports[`Fuzzy > should return correct results for a query 1`] = `
[
  {
    "distance": 0,
    "text": "apple",
  },
  {
    "distance": 4,
    "text": "pineapple",
  },
  {
    "distance": 5,
    "text": "orange",
  },
  {
    "distance": 5,
    "text": "banana",
  },
]
`;

exports[`Fuzzy > should return empty array for no matches 1`] = `[]`;

exports[`Fuzzy > should set this.list to an empty array if list is undefined 1`] = `
Fuzzy {
  "cache": {},
  "list": [],
  "options": {
    "includeMatches": false,
  },
  "search": [Function],
}
`;
